### ISSUE TITLE
.debug_names entry referring to skeleton type unit

### ISSUE THREAD

**Issue by dwblaikie** [llvm/llvm-project#77696]
When using `.debug_names` and Type Units, two entries for the type are produced ( https://godbolt.org/z/7Pd67afn7 ):
```
struct t1 { };
t1 v1;
```
```
     Hash: 0x59796A
      String: 0x00000085 "t1"
      Entry @ 0xa8 {
        Abbrev: 0x26c
        Tag: DW_TAG_structure_type
        DW_IDX_type_unit: 0x00
        DW_IDX_die_offset: 0x00000023
      }
      Entry @ 0xaf {
        Abbrev: 0x268
        Tag: DW_TAG_structure_type
        DW_IDX_die_offset: 0x0000002e
      }
```
```
DW_TAG_structure_type
                DW_AT_calling_convention	(DW_CC_pass_by_value)
                DW_AT_name	("t1")
                DW_AT_byte_size	(0x01)
                DW_AT_decl_file	("/app/example.cpp")
                DW_AT_decl_line	(1)
```
```
DW_TAG_structure_type
                DW_AT_declaration	(true)
                DW_AT_signature	(0xc6694e51369161f2)
```
Probably only want the type unit entry for the type in the index.

**Comment by clayborg** [comment]
We also need to verify that if a DIE has a `DW_IDX_parent` entry, that it doesn't add an entry to .debug_names that is a forward declaration. I am seeing this in a binary that has had BOLT run on it, so not sure if those entries are coming from clang or from BOLT.

Quoting from an email:
> fpiovezan@apple.com said:
>I need to find my notes from those days, but I don't think we did. As Greg points out, the standard forbids forward declarations in debug_names; entries whose parents are forward declarations should be marked as having a parent that is not indexed.
> The addition of IDX_parent_entries should not cause the addition of _entries_ in the table, if it does then it is a bug in the implementation.

> In fact, in the original PR we have this bit of code: https://github.com/llvm/llvm-project/pull/77457/files#diff-587587ad06ddb6f99f9ad8d8deffbc2ea59fde9d62d1b5ff58ace1f52cc75752R405

**Comment by dwblaikie** [comment]
> We also need to verify that if a DIE has a `DW_IDX_parent` entry, that it doesn't add an entry to .debug_names that is a forward declaration.

Probably best not to muddy other bugs like this - either file a separate one, or a new thread on discourse, etc. 

but here's the verification that we don't add a DW_IDX_parent for a declaration, at least one case of not doing that: https://godbolt.org/z/Ke3so5Eb5

**Comment by ayermolo** [comment]
This seems like a variation of another issue, where clang adds entries for DW_AT_declaration?

`
0x0000002e:   DW_TAG_structure_type
                DW_AT_declaration	(true)
                DW_AT_signature	(0xc6694e51369161f2)
`

**Comment by dwblaikie** [comment]
Fixed in bd5c6367bd79c01665ff8ab9848afbf5b1249ce6
